mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -124,6 +181,161 @@ impl PeerManager {
+                    .checked_add(Duration::from_millis(
+                        self.mempool_config.shared_mempool_ack_timeout_ms,
+                    ))
+                    .map_or(false, |deadline| {
+                        SystemTime::now().duration_since(deadline).is_ok()
+                    })
+                    || matches!(batch.ack_status, AckStatus::Retry)
+            })
+            .map(|(id, _batch)| (*id, mempool.timeline_range(id.0, id.1)));
+
+        let (batch_id, transactions) = match rebroadcast {
+            Some((batch_id, txns)) => {
+                debug!("rebroadcasting {:?}", batch_id);
+                (batch_id, txns)
+            }
+            None => {
+                // fresh broadcast
+                let (txns, new_timeline_id) = mempool.read_timeline(
+                    state.timeline_id,
+                    self.mempool_config.shared_mempool_batch_size,
+                );
+                (BatchId(state.timeline_id, new_timeline_id), txns)
+            }
+        };
+
+        if transactions.is_empty() {
+            return;
+        }
+
+        // don't hold mempool lock during network send
+        drop(mempool);
+
+        // execute actual network send
+        let mut network_sender = smp
+            .network_senders
+            .get_mut(&peer.network_id())
+            .expect("[shared mempool] missing network sender")
+            .clone();
+        let num_txns = transactions.len();
+        let send_time = SystemTime::now();
+        if let Err(e) = tasks::send_mempool_sync_msg(
+            MempoolSyncMsg::BroadcastTransactionsRequest {
+                request_id: batch_id.create_request_id(),
+                transactions,
+            },
+            peer.peer_id(),
+            &mut network_sender,
+        ) {
+            error!(
+                "[shared mempool] error broadcasting transactions to peer {:?}: {}",
+                peer, e
+            );
+        } else {
+            let peer_id = &peer.peer_id().to_string();
+            counters::SHARED_MEMPOOL_BROADCAST_LATENCY
+                .with_label_values(&[peer_id])
+                .observe(start_time.elapsed().as_secs_f64());
+            counters::SHARED_MEMPOOL_TRANSACTION_BROADCAST
+                .with_label_values(&[peer_id])
+                .observe(num_txns as f64);
+            counters::SHARED_MEMPOOL_PENDING_BROADCASTS_COUNT
+                .with_label_values(&[peer_id])
+                .inc();
+
+            // update peer sync state with info from above broadcast
+            state.timeline_id = std::cmp::max(state.timeline_id, batch_id.1);
+            // turn off backoff mode after every broadcast
+            state.broadcast_info.backoff_mode = false;
+            state.broadcast_info.sent_batches.insert(
+                batch_id,
+                BatchInfo {
+                    send_time,
+                    ack_status: AckStatus::Pending,
+                },
+            );
```

Comment:
considering we're still holding the peer lock, should we move the latency metric to after this? or even at the end of notify_subscribers? On the other hand, if we just put it at the beginning of the function using the start_timer, it would actually cover this entire function.